express-notes-api/index.js

```diff
@@ -0,0 +1,101 @@
+const express = require('express');
+const fs = require('fs');
+const app = express();
+
+require('./data.json');
+
+app.get('/api/notes/', (req, res) => {
+
+  const notesArray = [];
+
+  fs.readFile('./data.json', 'utf8', (err, data) => {
+    if (err) {
+      res.status(500).json('unexpected error');
+    }
+    const newJSON = JSON.parse(data);
+    for (const note in newJSON.notes) {
+      notesArray.push(newJSON.notes[note]);
+    }
+    res.status(200).json(notesArray);
+  });
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  if (req.params.id > 0) {
+    fs.readFile('./data.json', 'utf8', (err, data) => {
+      if (err) {
+        res.status(500).json('unexpected error');
+      }
+      const newJSON = JSON.parse(data);
+
+      if (newJSON.notes[req.params.id]) {
+        res.status(200).json(newJSON.notes[req.params.id]);
+      } else {
+        res.status(400).json(`Could not find a note with id ${req.params.id}`);
+      }
+    });
+  } else {
+    res.status(200).json('Invalid id');
+  }
+});
+
+app.delete('/api/notes/:id', (req, res) => {
+  if (req.params.id) {
+    fs.readFile('./data.json', 'utf8', (err, data) => {
+      if (err) {
+        res.status(500).json('unexpected error');
+      }
+      const newJSON = JSON.parse(data);
+
+      if (newJSON.notes[req.params.id]) {
+        delete newJSON.notes[req.params.id];
+        writeNote(newJSON, error => {
+          if (error) {
+            return res.status(500).json({ error: 'An unexpected server error occurred' });
+          }
+        });
+        res.status(204).json();
```

Comment:
This line needs to be within the `writeNote` callback, you don't want to send back a response until everything is complete and you know it worked.